### PR TITLE
Pupil numbers performance update

### PR DESCRIPTION
**Pull request by nwarms**
[User Story 174255](https://dfe-gov-uk.visualstudio.com/Academies-and-Free-Schools-SIP/_workitems/edit/174255): Build: Improve performance of Academies in this Trust pupil numbers page
Improved the performance of the Pupil numbers page in the academies in trust section.

## Changes

- Add methods for fetching pupil numbers to the academies repository and service
- Updated the the Pupil Numbers page to use the Academies Service over the trust provider
- Created the AcademyPupilNumbers and AcademyPupilNumbersServiceModel classes
- Created and added unit tests for the code

## Checklist

- [x] Pull request attached to the appropriate user story in Azure DevOps
- [x] Release notes added to CHANGELOG.md
- [x] Testing complete - all manual and automated tests pass

### CODE REVIEW

File: tests/DfE.FindInformationAcademiesTrusts.Data.AcademiesDb.UnitTests/Repositories/AcademyRepositoryTests.cs [hunk 1]
```diff
@@ -323,4 +323,38 @@ public async Task GetUrnForSingleAcademyTrustAsync_should_set_singleAcademyTrust
...
+        _mockAcademiesDbContext.AddGiasGroupLinksForGiasEstablishmentsToGiasGroup(giasEstablishments, giasGroup);
+
+        var result = await _sut.GetAcademiesInTrustPupilNumbersAsync("1234");
+
+        result.Should()
+            .BeEquivalentTo(giasEstablishments,
+                options => options
+                    .WithAutoConversion()
+                    .ExcludingMissingMembers()
+            );
```
[the remark this answers is not in the document]
Comment (nwarms) [reply tests/DfE.FindInformationAcademiesTrusts.Data.AcademiesDb.UnitTests/Repositories/AcademyRepositoryTests.cs:352]: Couldn't find an easy way to do this with the object equivalence (had some issues as "The member paths must have the same parent.")  so just did some loops over the arrays. It works but if you know a better way then let me know

File: DfE.FindInformationAcademiesTrusts.Data/Repositories/Academy/AcademyPupilNumbers.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+namespace DfE.FindInformationAcademiesTrusts.Data.Repositories.Academy;
+
+public record AcademyPupilNumbers(
+    string Urn,
+    string? EstablishmentName,
+    string? PhaseOfEducation,
+    AgeRange? AgeRange,
```
Comment (dynamictulip) [inline DfE.FindInformationAcademiesTrusts.Data/Repositories/Academy/AcademyPupilNumbers.cs:7]: The repository always creates one of these so this shouldn't be nullable

Comment (dynamictulip) [reply DfE.FindInformationAcademiesTrusts.Data/Repositories/Academy/AcademyPupilNumbers.cs:7]: then none of the places using it need to be nullable/have null access

File: tests/DfE.FindInformationAcademiesTrusts.Data.AcademiesDb.UnitTests/Repositories/AcademyRepositoryTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dynamictulip) [reply tests/DfE.FindInformationAcademiesTrusts.Data.AcademiesDb.UnitTests/Repositories/AcademyRepositoryTests.cs:352]: We'd have a separate test which has a couple of different age ranges and just checks that each age range is what it should be

File: tests/DfE.FindInformationAcademiesTrusts.UnitTests/Services/AcademyServiceTests.cs [hunk 3]
```diff
@@ -60,4 +61,37 @@ public async Task GetAcademiesInTrustOfstedAsync_should_return_mapped_result_fro
...
     }
+
+    [Fact]
+    public async Task GetAcademiesInTrustPupilNumbersAsync_should_return_mapped_result_from_repository()
+    {
+        const string uid = "1234";
+        AcademyPupilNumbers[] academies =
+        [
+            BuildDummyAcademyPupilNumbers("9876"),
+            BuildDummyAcademyPupilNumbers("9876")
```
Comment (dynamictulip) [inline tests/DfE.FindInformationAcademiesTrusts.UnitTests/Services/AcademyServiceTests.cs:72]: These two academies are identical so this test doesn't really test that things are mapped correctly, just that you get this particular result back. I'd make the two academies have every field different

File: DfE.FindInformationAcademiesTrusts/Pages/Trusts/Academies/PupilNumbers.cshtml [hunk 4]
```diff
@@ -18,15 +18,15 @@
...
         <tr class="govuk-table__row" data-testid="academy-row">
-            <th scope="row" class="govuk-body govuk-table__header" data-sort-value="@academy.EstablishmentName" data-testid="school-name">
+          <th scope="row" class="govuk-body govuk-table__header" data-sort-value="@academy.EstablishmentName" data-testid="school-name">
             @academy.EstablishmentName<br/><span class="govuk-!-font-weight-regular" data-testid="urn">URN: @academy.Urn</span>
           </th>
-          <td class="govuk-body govuk-table__cell" data-sort-value="@Model.PhaseAndAgeRangeSortValue(academy)" data-testid="phase-and-age-range">
+          <td class="govuk-body govuk-table__cell" data-sort-value="@PupilNumbersModel.PhaseAndAgeRangeSortValue(academy)" data-testid="phase-and-age-range">
             @academy.PhaseOfEducation<br/>
-            @academy.AgeRange.Minimum - @academy.AgeRange.Maximum
+            @academy.AgeRange?.Minimum - @academy.AgeRange?.Maximum
```
Comment (dynamictulip) [inline DfE.FindInformationAcademiesTrusts/Pages/Trusts/Academies/PupilNumbers.cshtml:29]:
```suggestion
            @academy.AgeRange.Minimum - @academy.AgeRange.Maximum
```

File: DfE.FindInformationAcademiesTrusts/Pages/Trusts/Academies/PupilNumbers.cshtml.cs [hunk 5]
```diff
@@ -33,8 +37,8 @@ public override async Task<IActionResult> OnGetAsync()
 
     public string TabName => "Pupil numbers";
 
-    public string PhaseAndAgeRangeSortValue(Academy academy)
+    public static string PhaseAndAgeRangeSortValue(AcademyPupilNumbersServiceModel academy)
     {
-        return $"{academy.PhaseOfEducation}{academy.AgeRange.Minimum:D2}{academy.AgeRange.Maximum:D2}";
+        return $"{academy.PhaseOfEducation}{academy.AgeRange?.Minimum:D2}{academy.AgeRange?.Maximum:D2}";
```
Comment (dynamictulip) [inline DfE.FindInformationAcademiesTrusts/Pages/Trusts/Academies/PupilNumbers.cshtml.cs:42]:
```suggestion
        return $"{academy.PhaseOfEducation}{academy.AgeRange.Minimum:D2}{academy.AgeRange.Maximum:D2}";
```

**Review by dynamictulip — APPROVED** [review]

**Review by dynamictulip — APPROVED** [review]
Prod code approved!

**Pull request merged**